### PR TITLE
Release resource of FI after all drivers have been closed

### PR DESCRIPTION
**Pull request by lancelly**
We can release resource of FI until all drivers have been closed, otherwise there could be memory leak/file handle leak.

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/fragment/FragmentInstanceContext.java [hunk 1]
```diff
@@ -352,6 +355,26 @@ private void addFilePathToMap(TsFileResource tsFile, boolean isClosed) {
     }
   }
 
+  public void initializeNumOfDrivers(Integer numOfDrivers) {
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/fragment/FragmentInstanceContext.java:358]:
```suggestion
  public void initializeNumOfDrivers(int numOfDrivers) {
```

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/fragment/FragmentInstanceContext.java [hunk 2]
```diff
@@ -352,6 +355,26 @@ private void addFilePathToMap(TsFileResource tsFile, boolean isClosed) {
...
+  }
+
+  public void releaseResourceWhenAllDriversAreClosed() {
+    try {
+      allDriversClosed.await();
+    } catch (InterruptedException e) {
+      LOGGER.warn(
+          "Interrupted when await on allDriversClosed, FragmentInstance Id is {}", this.getId());
+      throw new RuntimeException(e);
+    }
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/fragment/FragmentInstanceContext.java:374]:
```suggestion
    while (true) {
       try {
          allDriversClosed.await();
          break;
        } catch (InterruptedException e) {
          LOGGER.warn(
              "Interrupted when await on allDriversClosed, FragmentInstance Id is {}", this.getId());
        }
    }
```

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/fragment/FragmentInstanceContext.java (hunk 2 above)
Comment (lancelly) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/fragment/FragmentInstanceContext.java]: done

**Review by JackieTien97 — APPROVED** [review]

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/fragment/FragmentInstanceContext.java (hunk 1 above)
Comment (lancelly) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/fragment/FragmentInstanceContext.java]: done

**Pull request merged**